dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data.join;
+
+import com.google.common.collect.LinkedListMultimap;
+import com.google.common.collect.Multimap;
+import io.crate.data.ArrayRow;
+import io.crate.data.BatchIterator;
+import io.crate.data.Row;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * <pre>
+ *     Build Phase:
+ *     for (leftRow in left) {
+ *         calculate hash and put in Buffer (HashMap)
+ *     }
+ *     // All rows from left processed and inserted in the Buffer
+ *
+ *     Probe Phase:
+ *     // From now on we only iterate on the right
+ *     for (rightRow in right) {
+ *         if (hash(rightRow) found in Buffer {
+ *            for (row in matchedInBuffer) { // Handle duplicate values from left and hash-conflicts
+ *                if (joinCondition matches) {
+ *                    // We need to check that the joinCondition matches as we can have a hash-conflict
+ *                    // or the join condition can contain more operators. See HashJoinConditionSymbolsExtractor.
+ *                    emmit(combinedRow)
+ *                }
+ *            }
+ *         }
+ *     }
+ * </pre>
+ */
+class HashInnerJoinBatchIterator<L extends Row, R extends Row, C> extends JoinBatchIterator<L, R, C> {
+
+    private final Predicate<C> joinCondition;
+    private final Multimap<Integer, Object[]> buffer;
```

Comment:
One option to re-use the bucketing / collision logic of HashMap or other smarter Map implementations would be to parameterize this with a `L -> StoreType` and `R -> StoreType` function and rely on `hashCode` on StoreType being sensible. For now `StoreType` could be `List<Object>` 

Though we can also try this later on once we have some initial runtime stats

Reply:
Thx for the suggestion, but I think we postpone it for the moment and revisit when we test other hash functions.